### PR TITLE
Add comprehensive path normalization for artifact paths

### LINKED ISSUE #79: Comprehensive path normalization: handle leading/trailing slashes, prevent traversal
## Problem

Path handling is inconsistent across commands. Leading and trailing slashes cause errors or unexpected behavior instead of being normalized.

### Examples

**Push with leading slash fails:**
```
$ magpie push /tmp/file.txt --to /test/artifact
Error: Upload failed (400): None

$ magpie push /tmp/file.txt --to test/artifact
Uploaded: a6482406...  # Works
```

**LS with trailing slash fails:**
```
$ magpie ls test/
Error: Empty component in path 'test/' (double slash?)

$ magpie ls test/myartifact
HASH         TAGS...  # Works
```

**LS prefix listing fails:**
```
$ magpie ls test
Error: List failed (404): Not Found  # Should list artifacts under test/
```

### Path Normalization

All commands should normalize paths consistently:

| Input | Normalized | Notes |
|-------|------------|-------|
| `/test/artifact` | `test/artifact` | Strip leading slash |
| `test/artifact/` | `test/artifact` | Strip trailing slash |
| `//test//artifact//` | `test/artifact` | Collapse multiple slashes |
| `test/artifact` | `test/artifact` | Already clean |

### Path Traversal Prevention

Must reject:
- `../` sequences
- Paths that would escape storage root
- Paths with encoded traversal (`%2e%2e/`)

### Affected Commands

- `push --to PATH`
- `get PATH`
- `ls PATH` (both artifact listing and prefix listing)
- `info PATH`
- `tag PATH`
- `untag PATH`
- `url PATH`
- `amend PATH`

### Server-Side Validation

The server should also normalize and validate paths, not just the CLI. Defense in depth.

## Proposed Implementation

1. **Shared path normalization function** in `cli/` or `storage/paths.py`:
   ```python
   def normalize_artifact_path(path: str) -> str:
       """Normalize artifact path, stripping leading/trailing slashes."""
       # Strip leading/trailing slashes
       path = path.strip("/")
       # Collapse multiple slashes
       path = re.sub(r'/+', '/', path)
       # Reject traversal
       if ".." in path:
           raise ValueError("Path traversal not allowed")
       return path
   ```

2. **Apply in CLI** before sending to server
3. **Apply in server routes** as defense in depth
4. **Clear error messages** when paths are invalid

## Supersedes

This issue supersedes #75 (ls trailing slash) with a more comprehensive fix.

---
*Filed with AI assistance (Claude Code w/ Opus 4.5).*

### PR DESCRIPTION
**Pull request by duplico**
## Summary

- Add `normalize_artifact_path()` function in storage/paths.py for consistent path handling
- Apply normalization in all CLI commands via the parse module
- Apply normalization in server routes as defense in depth
- Add comprehensive unit tests for the normalization function

Fixes #79

## Test plan

- [x] Run `uv run pytest tests/unit tests/integration -x`
- [x] Verify `/test/artifact` normalizes to `test/artifact`
- [x] Verify `test/artifact/` normalizes to `test/artifact`
- [x] Verify `//test//artifact//` normalizes to `test/artifact`
- [x] Verify `test/../other` is rejected with path traversal error
- [x] Verify empty path is rejected

Note: Some pre-existing test failures unrelated to this PR (test_*_requires_server tests) are excluded from the test run.

---
Generated with [Claude Code](https://claude.ai/code) (Claude Opus 4.5)

### CODE REVIEW

File: src/magpie/storage/paths.py [hunk 1]
```diff
@@ -2,10 +2,58 @@
...
+        Raises InvalidArtifactPathError
+    """
+    # Strip leading and trailing slashes
+    path = path.strip("/")
+
+    # Collapse multiple consecutive slashes to single slash
+    path = re.sub(r"/+", "/", path)
+
+    # Check for path traversal attempts
+    if ".." in path:
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/storage/paths.py]: We should also consider an explicit integrity check that verifies, after constructing a path, it is strictly a descendent of the artifacts directory. This avoids entirely the potentially fragile searching for specific tokens approach.

[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/storage/paths.py]: Fixed in commit e2ac528. Changed from substring check (`if ".." in path`) to segment-based validation (`if any(segment == ".." for segment in segments)`), consistent with how `validate_artifact_path()` handles it at line 146. This correctly allows paths like "v1..2" or "test..file" while still blocking actual path traversal attempts.

Added tests to verify:
- `test_double_dot_in_filename_allowed`: Confirms "v1..2/artifact", "test..file", "project/name..ext" are accepted
- `test_double_dot_as_segment_rejected`: Confirms "test/../other", "../test", "test/.." are rejected

Generated with Claude Code (Claude Opus 4.5)

[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/storage/paths.py]: Added in commit e2ac528. I've implemented `verify_path_is_descendant()` which provides the explicit integrity check you suggested. It:

1. Constructs the full path by joining base and artifact path
2. Resolves both paths to canonical form
3. Verifies the resolved path is under the base directory using `relative_to()`
4. Rejects paths that resolve to the base directory itself

This complements the segment-based token validation with a belt-and-suspenders approach - even if the token check somehow missed something, the resolution check would catch any escape attempts.

Added comprehensive tests in `TestVerifyPathIsDescendant` including escape attempts, complex traversal, and edge cases.

Generated with Claude Code (Claude Opus 4.5)

File: tests/unit/test_path_validation.py [hunk 2]
```diff
@@ -181,3 +186,152 @@ def test_complex_nesting_scenario(self, tmp_path: Path) -> None:
...
+        result = verify_path_is_descendant(tmp_path, "v1..2/artifact")
+        assert result == tmp_path / "v1..2" / "artifact"
+
+    def test_path_stays_within_base(self, tmp_path: Path) -> None:
+        """Path that goes up then down but stays in base is valid."""
+        # Note: This tests that "project/../other" resolves to "other" which is valid
+        # The segment-based check in normalize_artifact_path would catch ".."
+        # but this test verifies the resolution behavior
+        result = verify_path_is_descendant(tmp_path, "project/../other")
+        assert result == tmp_path / "other"
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/unit/test_path_validation.py]: Fixed in f487509. Removed this test since it tested unreachable behavior - paths with "..` segments are rejected by `normalize_artifact_path()` before they ever reach `verify_path_is_descendant()`. The test was documenting behavior that could not occur in practice given the current architecture.

Generated with Claude Code (Claude Opus 4.5)

File: src/magpie/storage/paths.py [hunk 3]
```diff
@@ -2,10 +2,104 @@
...
+    Raises:
+        InvalidArtifactPathError: If the resolved path escapes the base directory.
+
+    Example:
+        >>> base = Path("/storage/artifacts")
+        >>> verify_path_is_descendant(base, "project/artifact")
+        PosixPath('/storage/artifacts/project/artifact')
+        >>> verify_path_is_descendant(base, "../escape")
+        Raises InvalidArtifactPathError
+    """
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/magpie/storage/paths.py:87]: Fixed in f487509. Integrated `verify_path_is_descendant()` into `StorageService.store_artifact()` to provide defense-in-depth path validation. Now paths are validated both by:
1. `validate_artifact_path()` - token-based validation for reserved names and hidden segments
2. `verify_path_is_descendant()` - resolution-based validation to ensure path stays within storage directory

Also updated the docstring to clarify that this function expects pre-normalized paths (those that have passed through `normalize_artifact_path()`), making its role as an additional safety layer explicit.

Generated with Claude Code (Claude Opus 4.5)
